src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,114 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final double LOAD_FACTOR = 0.75;
+    private Node<K, V>[] buckets;
+    private int size;
+
+    public MyHashMap() {
+        buckets = createBucketArray(DEFAULT_CAPACITY);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Node<K, V>[] createBucketArray(int capacity) {
+        return (Node<K, V>[]) new Node[capacity];
+    }
 
     @Override
     public void put(K key, V value) {
+        resizeIfNeeded();
+        int index = determineIndex(key);
+
+        // Handle the case where key is null and should be stored in bucket[0]
+        if (key == null) {
+            if (buckets[0] == null) {
+                buckets[0] = new Node<>(null, value);
+                size++;
+            } else {
+                insertOrUpdateInBucket(buckets[0], key, value);
+            }
+            return;
+        }
 
+        // For non-null keys, proceed with the determined index
+        if (buckets[index] == null) {
+            buckets[index] = new Node<>(key, value);
+            size++;
+        } else {
+            insertOrUpdateInBucket(buckets[index], key, value);
+        }
+    }
+
+    private void insertOrUpdateInBucket(Node<K, V> head, K key, V value) {
+        Node<K, V> current = head;
+
+        // Traverse the linked list to find either the key or the end
+        while (current != null) {
+            if (current.key == key || (current.key != null && current.key.equals(key))) {
+                current.value = value; // Update existing key
+                return;
+            }
+            if (current.next == null) {
+                current.next = new Node<>(key, value); // Append new node
+                size++;
```

Comment:
In the `insertOrUpdateInBucket` method, the `size` is incremented when a new node is appended. Ensure that the `size` is only incremented when a new key-value pair is added, not when an existing key is updated.